### ISSUE TITLE
fheroes2 crashed after moving any hero to the castle

### ISSUE THREAD

**Issue by darkhorror1984** [ihhub/fheroes2#5569]
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Windows

### Describe the bug

Game crashed after moving any hero to the castle. Map name's - ARRAX.

### Save file

[AUTOSAVE.zip](https://github.com/ihhub/fheroes2/files/8981696/AUTOSAVE.zip)

**Comment by idshibanov** [comment]
@darkhorror1984 what's the game version? Can't reproduce on latest build.

**Comment by oleg-derevenetz** [comment]
Hi @idshibanov this savefile was produced by 0.9.16, which is not even the latest release (which is 0.9.17), not to mention the latest build. However even 0.9.16 should not crash when visiting a castle.

Hi @darkhorror1984 could you please grab the output from the console window at the moment of the crash? If this crash is due to the assertion failure (and most likely it is) there may be some useful information - which assertion was failed and where.

**Comment by oleg-derevenetz** [comment]
Hmm, I still not able to reproduce the crash even with Russian resource files from Buka.

**Comment by darkhorror1984** [comment]
The console window is clear.
![crash_windows8](https://user-images.githubusercontent.com/108156692/175761786-a24f900e-3d3d-4cae-93fd-1a1e329d7b3a.png)

**Comment by oleg-derevenetz** [comment]
Hi @darkhorror1984 Could you please run the Event Viewer (by pressing `Win+R` and entering `eventvwr`), then open Windows Logs -> Application and look for related entries with source "Application Error" and "Windows Error Reporting"? If there are any, please copy their contents and paste it here. If there are links to some files with additional information, it would be good to get these files as a ZIP archive here as well.

**Comment by darkhorror1984** [comment]
[error_report.zip](https://github.com/ihhub/fheroes2/files/8984815/error_report.zip)

**Comment by darkhorror1984** [comment]
[error_report.zip](https://github.com/ihhub/fheroes2/files/8984867/error_report.zip)
[full_dump.zip](https://github.com/ihhub/fheroes2/files/8984868/full_dump.zip)

**Comment by oleg-derevenetz** [comment]
@darkhorror1984 Thank you! Interesting, the crash is inside SDL Mixer. Are you using MIDI music or external music (OGG/MP3)? What happens if you change the music type in game settings to the different one? Is the crash still observed after that?

**Comment by oleg-derevenetz** [comment]
Hi @ihhub since you have Windows PC with Visual Studio installed, could you please take a look on this dump? This dump was generated from the 0.9.16 release version I believe.

**Comment by darkhorror1984** [comment]
> @darkhorror1984 Thank you! Interesting, the crash is inside SDL Mixer. Are you using MIDI music or external music (OGG/MP3)? What happens if you change the music type in game settings to the different one? Is the crash still observed after that?
Music ans sound are off. I try music on and game isn't crashing now.

**Comment by oleg-derevenetz** [comment]
@darkhorror1984 Thus, if you turn off the music and sounds in the settings completely, then the game crashes when you enter the castle, however, if you turn on the music in the settings, then everything works just fine?

**Comment by darkhorror1984** [comment]
yes, everything works fine if i turn on music) I am use external music.

**Comment by oleg-derevenetz** [comment]
I haven't been able to reproduce the problem yet (even with the sound and music turned off), but I looked at the crash dump (@darkhorror1984 thank you very much for it). I don't have MSVC, all I have is the crash dump itself, so I personally didn't squeeze much out of it, but it looks like this crash happens in the separate thread apparently started by `SDL_Mixer`, probably to handle the playback of a music track. The crash is due to `nullptr` access (with offset) in this thread. Both `fheroes2` worker threads are waiting on their condition variables at that time, so it seems that music was successfully started by the worker thread, then `SDL_Mixer` starts an auxiliary thread for the castle music playback, and then it crashes due to the `nullptr` access. May be @ihhub will be able to take a closer look because he has the better toolset than me.

**Comment by ihhub** [comment]
I am unable to get any meaningful info but I compiled SDL2 0.9.16 debug version [fheroes2_debug.zip](https://github.com/ihhub/fheroes2/files/8986120/fheroes2_debug.zip)

@darkhorror1984 , can you please run this application for your save file? It will produce meaningful information for us.

**Comment by darkhorror1984** [comment]
I can't find any log files after crash.
![crash_debug](https://user-images.githubusercontent.com/108156692/175814768-3160fd24-a93b-43e9-9d4d-ce9cb3511ac8.png)

**Comment by oleg-derevenetz** [comment]
Hi @darkhorror1984 do you have crash dump left from the debug version? It would be good to attach it too. BTW you in fact don't use external music, because the game is unable to find it. You use MIDI music.

**Comment by darkhorror1984** [comment]
[Crash_localdump.zip](https://disk.yandex.ru/d/cvvrNCrlmvWC8A)

**Comment by ihhub** [comment]
The crash happens due to access out of memory boundaries. The crash dump didn't help. To fully solve the problem we have to build debug version of SDL...

**Comment by ihhub** [comment]
@darkhorror1984 , does it crash only for this particular scenario? What if you just open a castle without moving a hero there?

**Comment by darkhorror1984** [comment]
Thank you @asdf21
Solution:
1) Download http://www.libsdl.org/projects/mixer/timidity/timidity.tar.gz
2) Unpack them in /usr/local/lib under UNIX, and C:\ under Win32.

**Comment by oleg-derevenetz** [comment]
What a funny issue. Usually there is no need to use timidity on Windows, because Windows has its own built-in MIDI synthesizer with [API](https://docs.microsoft.com/en-us/windows/win32/multimedia/midi-functions) that is used by the `SDL_Mixer`.

**Comment by undef21** [comment]
@oleg-derevenetz, this issue reproduced only on win8 for midi music. External music works well.
I got this exception on same address as in @darkhorror1984 's crash dump but in the main menu. The crash in sdl_mixer
![windbg](https://user-images.githubusercontent.com/71143197/177872624-2b74418f-3e91-4fd4-9212-963e27f8aa62.png)

The workaround is above, but full investigation needs debug symbols for SDL and SDL_mixer or building debug versions of this libs. win8 will be end of life next half year so what about just adding this workaround to README?

**Comment by ihhub** [comment]
I believe this issue sits at the least on the edge of between our project, SDK and Windows 8. Adding something to README might not be a great solution as players won't look into it while facing the error.

I tried to use timidity on Windows 10: it fixes the issue with sound volume BUT the playback of sound track is wrong. I can hear incorrect instruments.

**Comment by ihhub** [comment]
The change to set timidity path will be added in the upcoming SDL Mixer 2.6 release.

**Comment by ihhub** [comment]
According to [release page](https://github.com/libsdl-org/SDL_mixer/releases) of SDL mixer they claim to "Fixed possible crash in win32 native midi". I recommend to test the newest version of SDL Mixer without timidity to see if the issue still persists.

**Comment by undef21** [comment]
> Adding something to README might not be a great solution as players won't look into it while facing the error.

We can add extraction of timidity.tar.gz to windows install script (only for win8).

> According to [release page](https://github.com/libsdl-org/SDL_mixer/releases) of SDL mixer they claim to "Fixed possible crash in win32 native midi". I recommend to test the newest version of SDL Mixer without timidity to see if the issue still persists.

This version doesn't export Mix_PlayChannel symbol (probably it's a bug). After manual building this lib the game crashes in the channelFinished callback.
![111](https://user-images.githubusercontent.com/71143197/178025091-c0258be6-9a60-4de0-9eef-b72056997050.png)

**Comment by undef21** [comment]
Looks like sdl_mixer team fixed export in 2.6.0 release. But the game crashes because we have recursion in channelFinished.

**Comment by oleg-derevenetz** [comment]
@undef21 We need it to properly cleanup the channel and free the memory chunk that was allocated for the sound being played. Apparently one of `SDL_Mixer` functions called for this now calls the callback once again. This callback was once written in such a way that it calls `SDL_Mixer` functions directly, which is called a "bad idea" in the manual. I added TODO comment regarding this while working on the music continuation feature together with @ihhub, but the current implementation previously worked fine, so finding a solution to the problem with calling `SDL_Mixer` functions from the callback has been postponed for now.

**Comment by undef21** [comment]
> Apparently one of SDL_Mixer functions called for this now calls the callback once again.

Mix_FreeChunk calls channelFinished.

**Comment by oleg-derevenetz** [comment]
I give priority to this issue because IMHO we need to sort this out before the next release. Otherwise, those who build the game with the new version of the `SDL_Mixer` (its package on various OSes will be eventually updated to the new version within the next month) will be very unhappy.

**Comment by oleg-derevenetz** [comment]
Hi @undef21 could you please test the branch from #5608 with new `SDL_Mixer` on your machine? This should eliminate crashes due to recursive callbacks.

**Comment by undef21** [comment]
@oleg-derevenetz I'm going to test this branch with win8 and fresh sdl_mixer on Monday

**Comment by oleg-derevenetz** [comment]
Now #5608 uses `SDL_Mixer` 2.6.0 (the latest one) to build the package using MSVC, so anyone can download any of the `fheroes2_windows_*_SDL2` artifacts from this PR and test it. On Windows 10 it seems to work just fine.

**Comment by oleg-derevenetz** [comment]
@undef21 Thank you very much!

> In the new version this behavior was fixed so lib doesn't crash and doesn't play MIDI.

Well, at least it doesn't crash anymore, which is already a progress :) I have an [idea](https://github.com/ihhub/fheroes2/issues/5528#issuecomment-1179796028) to always ship for Windows the version of `SDL_Mixer` built with `FluidSynth` (and no native MIDI). This should fix numerous issues with audio volume on Windows.
